controllers/subscriptionsController/subscribeSingleApplication.js

```diff
@@ -47,89 +83,143 @@ const subscribeSingleApplication = async (req, res, next) => {
       subscriptionDate.getMonth() + totalPeriod
     );
 
-    //populate subscription History Data
-    const subscriptionHistoryData = {
+    isSubscribed = await SubscriptionsModel.findOne({
       applicationId: applicationId,
-      planId: planId,
-      period: `${totalPeriod} months`,
-      expiresOn: expiryDate,
-      subscribedOn: subscriptionDate,
-    };
+    }).sort({ createdAt: "desc" });
 
-    //save to subscription history
-    const subHistory = new SubscriptionHistoryModel(subscriptionHistoryData);
-    await subHistory.save();
+    console.log(isSubscribed);
+    //check for existing app subsription
+    if (isSubscribed) {
+      const plans = [
+        { plan: "logging", planName: "maxLogRetentionPeriod" },
+        { plan: "dailyLimits", planName: "maxRequestPerDay" },
+        { plan: "perMinuteLimits", planName: "maxRequestPerMin" },
+      ];
 
-    //create subscription data & properties objects
-    //logging object
-    const logging = {
-      value: plan.logging,
-      maxLogRetentionDays: plan.maxLogRetentionPeriod,
-      expiryDate,
-    };
+      plans.forEach((item) => {
+        const planProperty = item.plan;
+        const planName = item.planName;
 
-    //request per min object
-    const perMinuteLimits = {
-      maxRequestsPerMin: plan.maxRequestPerMin,
-      expiryDate,
-    };
+        isSubscribed[planProperty].find((currPlan) => {
+          if (parseInt(currPlan[planName]) === parseInt(plan[planName])) {
+            let oldDate = new Date(currPlan.expiryDate.toLocaleString());
+            let curDate = new Date();
+            let newDate;
+            if (oldDate.getDate() < curDate.getDate()) {
+              newDate = new Date(expiryDate.toLocaleString());
+            } else {
+              newDate = newDateUtil(currPlan["expiryDate"], totalPeriod);
+            }
+            //deactive all existing plans and set only new one to true
+            deactivateAll(planProperty);
+            currPlan[planName] = plan[planName];
+            currPlan["expiryDate"] = new Date(newDate);
+            currPlan["isActive"] = true;
+            return true;
+          } else {
+            currPlan[planName] = plan[planName];
+            currPlan["expiryDate"] = expiryDate;
+            currPlan["isActive"] = true;
+            isSubscribed[planProperty].push(currPlan);
+          }
+        });
+      });
+      await isSubscribed.save();
+      await addSubToHistory(
+        applicationId,
+        planId,
+        totalPeriod,
+        expiryDate,
+        subscriptionDate,
+        isSubscribed._id
+      );
+      return responseHandler(
+        res,
+        201,
+        isSubscribed,
+        "Application subscription upgrade successful"
+      );
+    } else {
+      //create subscription data & properties objects
+      //logging object
+      const logging = {
+        value: plan.logging,
+        maxLogRetentionDays: plan.maxLogRetentionPeriod,
```

Comment:
I think we should stick with one here. maxLogRetentionsPerDay maybe for consistency?

Reply:
fixed